cli, security: add support for configurable TLS cipher suites

souravcrl opened the pull request:
cli, security: add support for configurable TLS cipher suites

fixes #136999
Epic CRDB-45351

Release Note(security,ops): We will be providing a new cockroach start command
cli flag `tls-cipher-suites` which is a string of comma separated list of cipher
suites to be used for all incoming TLS connections to the node. For TLS 1.2,
this should strictly be a subset of suites defined in
security/tls_ciphersuites.go as RecommendedCipherSuites or OldCipherSuites. For
TLS 1.3, this should be configured to a subset of ciphers in
crypto/tls/cipher_suites.go. The flag will restrict TLS connections to the node
for all 3 types of connection(i.e. sql, rpc, http) where node acts as the server
for the connections and close non-conforming ones.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137663)
<!-- Reviewable:end -->

souravcrl committed:
cli, security: add support for configurable TLS cipher suites

fixes #136999
Epic CRDB-45351

Release Note(security,ops): We will be providing a new cockroach start command
cli flag `tls-cipher-suites` which is a string of comma separated list of cipher
suites to be used for all incoming TLS connections to the node. For TLS 1.2,
this should strictly be a subset of suites defined in
security/tls_ciphersuites.go as RecommendedCipherSuites or OldCipherSuites. For
TLS 1.3, this should be configured to a subset of ciphers in
crypto/tls/cipher_suites.go. The flag will restrict TLS connections to the node
for all 3 types of connection(i.e. sql, rpc, http) where node acts as the server
for the connections and close non-conforming ones.

pritesh-lahoti reviewed (approved):
> Reviewed 10 of 10 files at r1, 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137663)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cthumuluru-crdb)



<!-- Sent from Reviewable.io -->

souravcrl commented:
The approach we took here of manually waiting for connection state change is too expensive as we only need to perform the cipher restrict check during handshake which can be forced on the connection. Hence have gone ahead with https://github.com/cockroachdb/cockroach/pull/143554
